File: WMS.DALCore/Producto/clsLnProducto_marca.cs
using Microsoft.Data.SqlClient;
using System.Data;
using System.Diagnostics;
using WMS.EntityCore.Producto;
using Microsoft.Extensions.Configuration;
using WMS.EntityCore.Producto.ProductoSimple;
using WMS.EntityCore.Interface;
public class clsLnProducto_Marca
{
    private static readonly clsInsert Inserter = new clsInsert();
    private static readonly clsUpdate Updater = new clsUpdate();

    private static string GetConnectionString(IConfiguration config) =>
        config.GetConnectionString("CST") ?? throw new InvalidOperationException("Cadena de conexión 'CST' no está configurada.");

    private static SqlCommand CreateCommand(string sql, SqlConnection conn, SqlTransaction? tx = null) =>
        new SqlCommand(sql, conn, tx) { CommandType = CommandType.Text };
    public static void Cargar(ref clsBeProducto_marca entity, DataRow row)
    {
        entity.IdMarca = row.Field<int?>("IdMarca") ?? 0;
        entity.IdPropietario = row.Field<int?>("IdPropietario") ?? 0;
        entity.Nombre = row.Field<string>("nombre") ?? "";
        entity.Activo = row.Field<bool?>("activo") ?? false;
        entity.User_agr = row.Field<string>("user_agr") ?? "";
        entity.Fec_agr = row.Field<DateTime?>("fec_agr") ?? DateTime.UtcNow;
        entity.User_mod = row.Field<string>("user_mod") ?? "";
        entity.Fec_mod = row.Field<DateTime?>("fec_mod") ?? DateTime.UtcNow;
        entity.Codigo = row.Field<string>("codigo") ?? "";
    }

    private static void BindParams(SqlCommand cmd, clsBeProducto_marca e)
    {
        cmd.Parameters.AddWithValue("@IdMarca", e.IdMarca);
        cmd.Parameters.AddWithValue("@IdPropietario", e.IdPropietario);
        cmd.Parameters.AddWithValue("@nombre", e.Nombre);
        cmd.Parameters.AddWithValue("@activo", e.Activo);
        cmd.Parameters.AddWithValue("@user_agr", e.User_agr);
        cmd.Parameters.AddWithValue("@fec_agr", e.Fec_agr);
        cmd.Parameters.AddWithValue("@user_mod", e.User_mod);
        cmd.Parameters.AddWithValue("@fec_mod", e.Fec_mod);
        cmd.Parameters.AddWithValue("@codigo", e.Codigo);
    }

    public static int Insert(clsBeProducto_marca entity, SqlConnection connection, SqlTransaction transaction)
    {
        Inserter.Init("producto_marca");
        Inserter.Add("idmarca", "@idmarca", "F");
        Inserter.Add("idpropietario", "@idpropietario", "F");
        Inserter.Add("nombre", "@nombre", "F");
        Inserter.Add("activo", "@activo", "F");
        Inserter.Add("user_agr", "@user_agr", "F");
        Inserter.Add("fec_agr", "@fec_agr", "F");
        Inserter.Add("user_mod", "@user_mod", "F");
        Inserter.Add("fec_mod", "@fec_mod", "F");
        Inserter.Add("codigo", "@codigo", "F");

        string sql = Inserter.SQL();

        try
        {
            using var cmd = new SqlCommand(sql, connection, transaction);
            BindParams(cmd, entity);
            int result = cmd.ExecuteNonQuery();
            return result;
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int Update(clsBeProducto_marca entity, SqlConnection connection, SqlTransaction transaction)
    {
        Updater.Init("producto_marca");
        Updater.Add("idpropietario", "@idpropietario", "F");
        Updater.Add("nombre", "@nombre", "F");
        Updater.Add("activo", "@activo", "F");
        Updater.Add("user_agr", "@user_agr", "F");
        Updater.Add("fec_agr", "@fec_agr", "F");
        Updater.Add("user_mod", "@user_mod", "F");
        Updater.Add("fec_mod", "@fec_mod", "F");
        Updater.Add("codigo", "@codigo", "F");
        Updater.Where("idmarca = @idmarca");

        string sql = Updater.SQL();

        try
        {
            using var cmd = new SqlCommand(sql, connection, transaction);
            BindParams(cmd, entity);
            int result = cmd.ExecuteNonQuery();
            return result;
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int InsertOrUpdate(clsBeProducto_marca entity, SqlConnection connection, SqlTransaction transaction)
    {
        try
        {
            int result = Existe(entity.IdMarca, connection, transaction)
                ? Update(entity, connection, transaction)
                : Insert(entity, connection, transaction);

            return result;
        }
        catch
        {
            throw;
        }
    }
    public static bool Existe(int idMarca, SqlConnection conn, SqlTransaction tx)
    {
        try
        {
            const string sql = "SELECT COUNT(1) FROM Producto_marca WHERE IdMarca = @IdMarca";
            using var cmd = new SqlCommand(sql, conn, tx);
            cmd.Parameters.AddWithValue("@IdMarca", idMarca);
            return Convert.ToInt32(cmd.ExecuteScalar()) > 0;
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static List<clsBeProducto_marca> GetAll(IConfiguration config)
    {
        const string sql = "SELECT * FROM Producto_marca";
        using var connection = new SqlConnection(GetConnectionString(config));
        connection.Open();
        using var tx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);

        try
        {
            using var cmd = CreateCommand(sql, connection, tx);
            using var adapter = new SqlDataAdapter(cmd);

            var table = new DataTable();
            adapter.Fill(table);

            tx.Commit();

            var list = new List<clsBeProducto_marca>();
            foreach (DataRow row in table.Rows)
            {
                var item = new clsBeProducto_marca();
                Cargar(ref item, row);
                list.Add(item);
            }

            return list;
        }
        catch (Exception ex)
        {
            tx.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static clsBeProducto_marca? GetById(IConfiguration config, int idMarca)
    {
        const string sql = "SELECT * FROM Producto_marca WHERE IdMarca = @IdMarca";
        using var connection = new SqlConnection(GetConnectionString(config));
        connection.Open();
        using var tx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);

        try
        {
            using var cmd = CreateCommand(sql, connection, tx);
            cmd.Parameters.AddWithValue("@IdMarca", idMarca);

            using var adapter = new SqlDataAdapter(cmd);
            var table = new DataTable();
            adapter.Fill(table);

            tx.Commit();

            if (table.Rows.Count == 0) return null;

            var result = new clsBeProducto_marca();
            Cargar(ref result, table.Rows[0]);
            return result;
        }
        catch (Exception ex)
        {
            tx.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int MaxId(IConfiguration config)
    {
        const string sql = "SELECT ISNULL(MAX(IdMarca), 0) FROM Producto_marca";
        using var connection = new SqlConnection(GetConnectionString(config));
        connection.Open();
        using var tx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);

        try
        {
            using var cmd = CreateCommand(sql, connection, tx);
            var result = cmd.ExecuteScalar();
            tx.Commit();

            return result != null && result != DBNull.Value ? Convert.ToInt32(result) : 0;
        }
        catch (Exception ex)
        {
            tx.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }

    public static int MaxId(SqlConnection connection, SqlTransaction transaction)
    {
        const string sql = "SELECT ISNULL(MAX(IdMarca), 0) FROM Producto_marca";

        try
        {
            using var cmd = CreateCommand(sql, connection, transaction);
            var result = cmd.ExecuteScalar();
            return result != null && result != DBNull.Value ? Convert.ToInt32(result) : 0;
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }

    public static bool Existe_By_Codigo(string Codigo, ref clsBeProducto_marca pBeMarca ,SqlConnection cn, SqlTransaction? tx = null)
    {
        try
        {
            const string sql = "SELECT TOP 1 * FROM producto_marca WHERE codigo = @codigo";

            using var cmd = new SqlCommand(sql, cn, tx);
            cmd.Parameters.AddWithValue("@codigo", Codigo);


            using var da = new SqlDataAdapter(cmd);
            var dt = new DataTable();
            da.Fill(dt);

            if (dt.Rows.Count == 1)
            {
                Cargar(ref pBeMarca, dt.Rows[0]);
                return true;
            }

            return false;
            
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static void Valida_Atributos(clsBeProducto_marcaSimple entity, SqlConnection connection, SqlTransaction transaction)
    {
        try
        {
            var Marca = new clsBeProducto_marca();
            bool existe = Existe_By_Codigo(entity.Codigo, ref Marca, connection, transaction);

            var BeInavConfigEnc = new clsBeI_nav_config_enc();
            clsLnI_nav_config_enc.GetSingle(BeInavConfigEnc, connection, transaction);

            if (BeInavConfigEnc == null)
                throw new ArgumentNullException(nameof(BeInavConfigEnc), "No se encuentra interface para definir propiedades de auditoria.");

            if (!existe)
            {
                if (!string.IsNullOrEmpty(entity.Codigo))
                {
                    Marca.IdMarca = MaxId(connection, transaction) + 1;
                    Marca.Codigo = entity.Codigo;
                    Marca.Nombre = entity.Nombre ?? entity.Codigo;
                    Marca.User_agr = BeInavConfigEnc.IdUsuario.ToString();
                    Marca.User_mod = BeInavConfigEnc.IdUsuario.ToString();
                    Marca.Fec_agr = DateTime.Now;
                    Marca.Fec_mod = DateTime.Now;
                    Marca.Activo = entity.Activo;
                    Marca.IdPropietario = entity.IdPropietario;
                    Insert(Marca, connection, transaction);
                }
            }
            else
            {
                Marca.Codigo = entity.Codigo;
                Marca.Nombre = entity.Nombre ?? entity.Codigo;
                Marca.User_mod = BeInavConfigEnc.IdUsuario.ToString();
                Marca.Fec_mod = DateTime.Now;
                Marca.Activo = entity.Activo;
                Update(Marca, connection, transaction);
            }
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name}: {ex.Message}", ex);
        }
    }


}

File: TOMIMSV4/DAL/Mantenimientos/Producto/ProductoMarca/clsLnProducto_marca.vb
Imports System.Data.SqlClient

Public Class clsLnProducto_marca

    Public Shared Sub Cargar(ByRef oBeProducto_marca As clsBeProducto_marca, ByRef dr As DataRow)
        Try
            With oBeProducto_marca
                .IdMarca = IIf(IsDBNull(dr.Item("IdMarca")), 0, dr.Item("IdMarca"))
                .Codigo = IIf(IsDBNull(dr.Item("Codigo")), 0, dr.Item("Codigo"))
                .IdPropietario = IIf(IsDBNull(dr.Item("IdPropietario")), 0, dr.Item("IdPropietario"))
                .Nombre = IIf(IsDBNull(dr.Item("nombre")), "", dr.Item("nombre"))
                .Activo = IIf(IsDBNull(dr.Item("activo")), False, dr.Item("activo"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
                .User_mod = IIf(IsDBNull(dr.Item("user_mod")), "", dr.Item("user_mod"))
                .Fec_mod = IIf(IsDBNull(dr.Item("fec_mod")), Date.Now, dr.Item("fec_mod"))
            End With
        Catch ex As Exception
            Throw ex
        End Try
    End Sub

    Public Shared Function Insertar(ByRef oBeProducto_marca As clsBeProducto_marca, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            Ins.Init("producto_marca")
            Ins.Add("idmarca", "@idmarca", DataType.Parametro)
            Ins.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Ins.Add("codigo", "@codigo", DataType.Parametro)
            Ins.Add("nombre", "@nombre", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Ins.Add("user_mod", "@user_mod", DataType.Parametro)
            Ins.Add("fec_mod", "@fec_mod", DataType.Parametro)

            Dim sp As String = Ins.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text


            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDMARCA", oBeProducto_marca.IdMarca))
            cmd.Parameters.Add(New SqlParameter("@CODIGO", oBeProducto_marca.Codigo))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBeProducto_marca.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", clsPublic.Quitar_Caracteres_No_Permitidos(oBeProducto_marca.Nombre)))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_marca.Activo))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_marca.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_marca.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_marca.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_marca.Fec_mod))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then
                lTransaction.Commit()
            End If

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeProducto_marca As clsBeProducto_marca, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            Upd.Init("producto_marca")
            Upd.Add("idmarca", "@idmarca", DataType.Parametro)
            Upd.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Upd.Add("codigo", "@codigo", DataType.Parametro)
            Upd.Add("nombre", "@nombre", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Add("user_mod", "@user_mod", DataType.Parametro)
            Upd.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Upd.Where("IdMarca = @IdMarca")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text


            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If
            cmd.Parameters.Add(New SqlParameter("@IDMARCA", oBeProducto_marca.IdMarca))
            cmd.Parameters.Add(New SqlParameter("@CODIGO", oBeProducto_marca.Codigo))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBeProducto_marca.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", clsPublic.Quitar_Caracteres_No_Permitidos(oBeProducto_marca.Nombre)))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_marca.Activo))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_marca.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_marca.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_marca.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_marca.Fec_mod))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then
                lTransaction.Commit()
            End If

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Eliminar(ByRef oBeProducto_marca As clsBeProducto_marca, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            cmd.CommandType = CommandType.Text

            Dim sp As String = " Delete from Producto_marca" &
             "  Where(IdMarca = @IdMarca)"


            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDMARCA", oBeProducto_marca.IdMarca))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Function Listar() As DataTable

        Try

            Const sp As String = "SELECT * FROM Producto_marca"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable
            dad.Fill(dt)

            Return dt

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function Get_Single_By_Codigo(ByVal pCodigo As String, pConnection As SqlConnection, pTransaction As SqlTransaction) As clsBeProducto_marca


        Get_Single_By_Codigo = Nothing

        Try

            Dim IdxProductoClasificacion As Integer = 0

            Dim vSQL As String = "SELECT TOP 1 * FROM producto_marca WHERE Codigo = '" + pCodigo + "'"

            Using lDTA As New SqlDataAdapter(vSQL, pConnection)

                lDTA.SelectCommand.CommandType = CommandType.Text
                lDTA.SelectCommand.Transaction = pTransaction

                Dim lDT As New DataTable()
                lDTA.Fill(lDT)

                If lDT IsNot Nothing AndAlso lDT.Rows.Count > 0 Then

                    Dim lRow As DataRow = lDT.Rows(0)
                    Dim BeProductoMarca As New clsBeProducto_marca()
                    Cargar(BeProductoMarca, lRow)
                    Return BeProductoMarca

                End If

            End Using

        Catch ex As Exception
            Throw ex
        End Try



    End Function

    Public Shared Function Get_Single_By_Nombre_By_IdPropietario(ByVal pNombre As String,
                                                                 ByVal pIdPropietario As Integer,
                                                                 pConnection As SqlConnection,
                                                                 pTransaction As SqlTransaction) As clsBeProducto_marca


        Get_Single_By_Nombre_By_IdPropietario = Nothing

        Try

            Dim IdxProductoClasificacion As Integer = 0

            Dim vSQL As String = "SELECT TOP 1 * FROM producto_marca 
                                  WHERE IdPropietario=@IdPropietario Nombre like '%" + pNombre + "%'"

            Using lDTA As New SqlDataAdapter(vSQL, pConnection)

                lDTA.SelectCommand.CommandType = CommandType.Text
                lDTA.SelectCommand.Parameters.Add(New SqlParameter("@IdPropietario", pIdPropietario))
                lDTA.SelectCommand.Transaction = pTransaction

                Dim lDT As New DataTable()
                lDTA.Fill(lDT)

                If lDT IsNot Nothing AndAlso lDT.Rows.Count > 0 Then

                    Dim lRow As DataRow = lDT.Rows(0)
                    Dim Obj As New clsBeProducto_marca()
                    Cargar(Obj, lRow)

                    Get_Single_By_Nombre_By_IdPropietario = Obj

                    Return Get_Single_By_Nombre_By_IdPropietario

                End If

            End Using

        Catch ex As Exception
            Throw ex
        End Try



    End Function

    Public Shared Function MaxID(ByVal pConnection As SqlConnection, ByVal pTransaction As SqlTransaction) As Integer

        Try

            Dim lMax As Integer = 0

            Dim vSQL As String = "SELECT ISNULL(MAX(IdMarca),0) + 1 as nuevo FROM producto_marca"

            Using lCommand As New SqlCommand(vSQL, pConnection, pTransaction)
                lCommand.CommandType = CommandType.Text
                Dim lReturnValue As Object = lCommand.ExecuteScalar()
                If lReturnValue IsNot DBNull.Value AndAlso lReturnValue IsNot Nothing Then
                    lMax = CInt(lReturnValue)
                End If
                lCommand.Dispose()
            End Using

            Return lMax

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function Get_Single_By_Nombre(ByVal pCodigo As String, pConnection As SqlConnection, pTransaction As SqlTransaction) As clsBeProducto_marca


        Get_Single_By_Nombre = Nothing

        Try

            Dim IdxProductoClasificacion As Integer = 0

            Dim vSQL As String = "SELECT TOP 1 * FROM producto_marca WHERE Nombre like '" + pCodigo + "%'"

            Using lDTA As New SqlDataAdapter(vSQL, pConnection)

                lDTA.SelectCommand.CommandType = CommandType.Text
                lDTA.SelectCommand.Transaction = pTransaction

                Dim lDT As New DataTable()
                lDTA.Fill(lDT)

                If lDT IsNot Nothing AndAlso lDT.Rows.Count > 0 Then

                    Dim lRow As DataRow = lDT.Rows(0)
                    Dim Obj As New clsBeProducto_marca()
                    Cargar(Obj, lRow)
                    Return Obj

                End If

            End Using

        Catch ex As Exception
            Throw ex
        End Try



    End Function

End Class
